File: common/scala/src/main/scala/org/aiddl/common/scala/math/linear_algebra/LupInverter.scala
package org.aiddl.common.scala.math.linear_algebra

import scala.collection.mutable.HashMap

import org.aiddl.core.scala.function.Function
import org.aiddl.core.scala.representation._

import org.aiddl.common.scala.math.linear_algebra.Matrix
import org.aiddl.common.scala.math.linear_algebra.AiddlMatrix
import org.aiddl.common.scala.math.linear_algebra.LupDecomposition
import org.aiddl.common.scala.math.linear_algebra.LupSolver

import org.aiddl.core.scala.representation.TermImplicits.term2Num
import org.aiddl.core.scala.representation.TermImplicits.double2Num

class LupInverter extends Function {
  def apply( x: Term ): Term = ???

  val lupDecompose = new LupDecomposition
  val lupSolve = new LupSolver

  def apply( a: Matrix ): Matrix = {
    val (l, u, p) = lupDecompose(a)
    val identity = AiddlMatrix.identity(a.m)

    val a_inv = Tuple((0 until a.m).map( i => {
      val sol = lupSolve(l, u, p, identity.row(i))
      Tuple((0 until a.n).map(j => sol(j)): _*)
    }): _*)

    AiddlMatrix(a_inv).t
  }
}

File: common/java/src/main/java/org/aiddl/common/java/math/linear_algebra/LupInverter.java
package org.aiddl.common.java.math.linear_algebra;

import org.aiddl.core.java.interfaces.Function;
import org.aiddl.core.java.representation.Term;
import org.aiddl.core.java.tools.LockableList;

public class LupInverter implements Function {

	@Override
	public Term apply(Term A) {
	
		LupDecomposition lupDecomp = new LupDecomposition();
		
		Term LUP = lupDecomp.apply(A);
				
		Term L  = LUP.get(0);
		Term U  = LUP.get(1);
		Term pi = LUP.get(2);
		
		LupSolver lupSolver = new LupSolver();
		
		IdentityMatrix ident = new IdentityMatrix();
		Term I = ident.apply(Term.integer(A.size()));
		
		LockableList A_inv = new LockableList();
		
		for ( int i = 0 ; i < I.size() ; i++ ) {
			Term x_i = lupSolver.apply(Term.tuple(L, U, pi, I.get(i)));	
			A_inv.add(x_i);
		}
		
		return Term.tuple(A_inv);
	}

}
